driver/src/contracts/stablex_auction_element.rs

```diff
@@ -0,0 +1,313 @@
+}
+
+impl StableXAuctionElement {
+    pub fn in_auction(&self, index: U256) -> bool {
+        self.valid_from <= index && index <= self.valid_until
+    }
+
+    pub fn from_bytes(order_count: &mut HashMap<H160, u16>, bytes: &[u8; 113]) -> Self {
+        let account_id = H160::from(&bytes[0..20]);
+
+        // these go together (since sell_token_balance is emitted as u256 and treated as u128
+        let sell_token_balance = BigEndian::read_u128(&bytes[36..52]);
+        let sell_token_balance_padding = BigEndian::read_u128(&bytes[20..36]);
+        assert_eq!(
+            sell_token_balance_padding, 0,
+            "User has too large balance to handle."
+        );
+
+        let buy_token = u16::from_le_bytes([bytes[53], bytes[52]]);
+        let sell_token = u16::from_le_bytes([bytes[55], bytes[54]]);
+        let valid_from = U256::from(u32::from_le_bytes([
+            bytes[59], bytes[58], bytes[57], bytes[56],
+        ]));
+        let valid_until = U256::from(u32::from_le_bytes([
+            bytes[63], bytes[62], bytes[61], bytes[60],
+        ]));
+        let is_sell_order = bytes[64] > 0;
+        let numerator = BigEndian::read_u128(&bytes[65..81]);
+        let denominator = BigEndian::read_u128(&bytes[81..97]);
+        let amount = BigEndian::read_u128(&bytes[97..113]);
+        let (buy_amount, sell_amount) =
+            compute_buy_sell_amounts(numerator, denominator, amount, is_sell_order);
+        let order_counter = order_count.entry(account_id).or_insert(0);
+        *order_counter += 1;
+        StableXAuctionElement {
+            valid_from,
+            valid_until,
+            sell_token_balance,
+            order: Order {
+                batch_information: Some(BatchInformation {
+                    slot_index: *order_counter - 1,
+                    slot: U256::from(0),
+                }),
+                account_id,
+                buy_token,
+                sell_token,
+                buy_amount,
+                sell_amount,
+            },
+        }
+    }
+}
+
+fn compute_buy_sell_amounts(
+    numerator: u128,
+    denominator: u128,
+    remaining: u128,
+    is_sell_order: bool,
+) -> (u128, u128) {
+    // TODO - use u128_to_u256 for overflow handling (this is a conversion used in naive_solver)
+    assert!(
+        remaining <= denominator,
+        "Smart contract should never allow this inequality"
+    );
+    let mut other = 0 as u128;
+    let (buy_amount, sell_amount) = if is_sell_order {
+        // 0 on sellAmount (remaining <= denominator) is nonsense, but solver can handle it.
+        // 0 on buyAmount (numerator) is a Market Sell Order.
+        if denominator > 0 {
+            // up-casting to handle overflow
+            let top = u128_to_u256(remaining) * u128_to_u256(numerator);
+            let bottom = u128_to_u256(denominator);
+            // Recall that ceil(p / float(q)) == (p + q - 1) / q
+            let other_u256 = (top + bottom - 1) / bottom;
```

Comment:
The naming here is quite terrible. You can reuse variable names. E.g. this could be `let buy_amount`

Reply:
was not aware I could reuse variable names.